lib/charms/tls_certificates_interface/v4/tls_certificates.py

```diff
@@ -0,0 +1,1499 @@
+        except DataValidationError:
+            requirer_relation_data = RequirerData(
+                certificate_signing_requests=[],
+            )
+        new_relation_data = copy.deepcopy(requirer_relation_data.certificate_signing_requests)
+        new_relation_data.append(new_csr)
+        try:
+            RequirerData(certificate_signing_requests=new_relation_data).dump(
+                relation.data[app_or_unit]
+            )
+            logger.info("Certificate signing request added to relation data.")
+        except ModelError:
+            logger.warning("Failed to update relation data")
+
+    def _send_certificate_requests(self):
+        if not self.private_key:
+            logger.debug("Private key not generated yet.")
+            return
+        for certificate_request in self.certificate_requests:
+            if not certificate_request.is_valid():
+                logger.warning("Invalid certificate request - Skipping")
+                continue
+            if not self._certificate_requested(certificate_request):
+                csr = generate_csr(
+                    private_key=self.private_key.private_key.encode(),
+                    sans_dns=certificate_request.sans_dns,
+                    common_name=certificate_request.common_name,
+                    organization=certificate_request.organization,
+                    email_address=certificate_request.email_address,
+                    country_name=certificate_request.country_name,
+                    state_or_province_name=certificate_request.state_or_province_name,
+                    locality_name=certificate_request.locality_name,
+                )
+                self._request_certificate(csr=csr.decode(), is_ca=certificate_request.is_ca)
+
+    def get_assigned_certificate(
+        self, certificate_request: CertificateRequest
+    ) -> Tuple[ProviderCertificate | None, PrivateKey | None]:
+        """Get the certificate that was assigned to the given certificate request."""
+        if requirer_csr := self.get_certificate_signing_request(certificate_request):
+            return self._find_certificate_in_relation_data(requirer_csr.csr), self.private_key
+        return None, None
+
+    def get_assigned_certificates(self) -> Tuple[List[ProviderCertificate], PrivateKey | None]:
+        """Get a list of certificates that were assigned to this unit."""
+        assigned_certificates = []
+        for requirer_csr in self.get_certificate_signing_requests(fulfilled_only=True):
+            if cert := self._find_certificate_in_relation_data(requirer_csr.csr):
+                assigned_certificates.append(cert)
+        return assigned_certificates, self.private_key
+
+    def get_certificate_signing_request(
+        self, certificate_request: CertificateRequest
+    ) -> Optional[RequirerCSR]:
+        """Get the CSR that was sent to the provider for the given certificate request."""
```

Comment:
I'm not sure there is value in returning an object. If the conditions are true, shouldn't it be identical with the one passed as parameter? isn't it better if this function returned a boolean and was named something like `certificate_signing_request_sent`?

Reply:
There is a difference between `CertificateRequest` which is the object that the lib v4 takes as an input from the requiring charm and `RequirerCSR`, which is the object we return to the user when they retrieve the list of requests they made. Maybe we should combine those two, I'm not sure

Reply:
Might be relevant to https://github.com/canonical/tls-certificates-interface/pull/197#discussion_r1686887958

Reply:
Allright, I completely got rid of the requirerCSR class as well as the methods to retrieve those as users should not need to manage CSR's at all anymore.